File: backend/src/utils/videoCatalog.js
const fs = require('fs');
const path = require('path');

function listVideos(mediaRoot) {
  const hlsRoot = path.join(mediaRoot, 'hls');
  const spritesRoot = path.join(mediaRoot, 'sprites');
  if (!fs.existsSync(hlsRoot)) return [];

  const entries = fs.readdirSync(hlsRoot, { withFileTypes: true });
  const videos = [];

  for (const dirent of entries) {
    if (!dirent.isDirectory()) continue;
    const id = dirent.name;
    const masterPath = path.join(hlsRoot, id, 'master.m3u8');
    if (!fs.existsSync(masterPath)) continue;

    videos.push({
      id,
      title: id,
      masterUrl: `/hls/${id}/master.m3u8`,
      sprite: resolveSprite(spritesRoot, id),
    });
  }

  return videos;
}

// スプライトが単一シート (id.jpg) の場合も、複数シート (id-1.jpg, id-2.jpg, …) の
// 場合も同じ形で返す。フロントは sheets[] をタイル index から引いて切替表示する。
function resolveSprite(spritesRoot, id) {
  const spriteJson = path.join(spritesRoot, `${id}.json`);
  if (!fs.existsSync(spriteJson)) return null;

  let meta;
  try { meta = JSON.parse(fs.readFileSync(spriteJson, 'utf8')); }
  catch (_) { return null; }

  const sheets = [];
  const sheetCount = Math.max(1, meta.sheetCount || 1);

  if (sheetCount === 1) {
    const p = path.join(spritesRoot, `${id}.jpg`);
    if (fs.existsSync(p)) sheets.push(`/sprites/${id}.jpg`);
  } else {
    for (let i = 1; i <= sheetCount; i++) {
      const p = path.join(spritesRoot, `${id}-${i}.jpg`);
      if (fs.existsSync(p)) sheets.push(`/sprites/${id}-${i}.jpg`);
    }
  }

  if (sheets.length === 0) return null;

  const spriteVtt = path.join(spritesRoot, `${id}.vtt`);
  return {
    sheets,
    sheetCount: sheets.length,
    vttUrl: fs.existsSync(spriteVtt) ? `/sprites/${id}.vtt` : null,
    tileWidth: meta.tileWidth,
    tileHeight: meta.tileHeight,
    columns: meta.columns,
    rows: meta.rows || 10,
    interval: meta.interval,
    tileCount: meta.tileCount,
  };
}

module.exports = { listVideos, resolveSprite };

File: hls_video/video_catalog.py
"""変換済み動画（master.m3u8 を持つ）の一覧取得とスプライト情報の解決。

Node 側 utils/videoCatalog.js と等価。
UI は常に `sheets[]` 配列を通じてスプライトシートを参照する。
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Optional


def list_videos(media_root: str) -> list[dict]:
    hls_root = Path(media_root) / "hls"
    sprites_root = Path(media_root) / "sprites"
    if not hls_root.exists():
        return []

    videos: list[dict] = []
    for entry in sorted(hls_root.iterdir()):
        if not entry.is_dir():
            continue
        master = entry / "master.m3u8"
        if not master.exists():
            continue
        video_id = entry.name
        videos.append({
            "id": video_id,
            "title": video_id,
            "master_url": f"/hls/{video_id}/master.m3u8",
            "sprite": resolve_sprite(str(sprites_root), video_id),
        })
    return videos


def resolve_sprite(sprites_root: str, video_id: str) -> Optional[dict]:
    meta_path = Path(sprites_root) / f"{video_id}.json"
    if not meta_path.exists():
        return None
    try:
        meta = json.loads(meta_path.read_text())
    except Exception:
        return None

    sheet_count = max(1, int(meta.get("sheetCount") or 1))
    sheets: list[str] = []
    if sheet_count == 1:
        if (Path(sprites_root) / f"{video_id}.jpg").exists():
            sheets.append(f"/sprites/{video_id}.jpg")
    else:
        for i in range(1, sheet_count + 1):
            if (Path(sprites_root) / f"{video_id}-{i}.jpg").exists():
                sheets.append(f"/sprites/{video_id}-{i}.jpg")

    if not sheets:
        return None

    vtt_path = Path(sprites_root) / f"{video_id}.vtt"
    return {
        "sheets": sheets,
        "sheet_count": len(sheets),
        "vtt_url": f"/sprites/{video_id}.vtt" if vtt_path.exists() else None,
        "tile_width": meta.get("tileWidth"),
        "tile_height": meta.get("tileHeight"),
        "columns": meta.get("columns"),
        "rows": meta.get("rows", 10),
        "interval": meta.get("interval"),
        "tile_count": meta.get("tileCount"),
    }
